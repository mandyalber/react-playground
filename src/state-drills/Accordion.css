.accordion {
    width: 100%;
    margin: 0;
    padding: 0;
    list-style: none;
}

.section {
    margin-bottom: 20px;
}

button {
    padding: 10px;
    border-radius: 5px;
    width: 100%;
    font-family: 'Roboto', sans-serif;
    font-size: 16px;
}

.content {
    background-color: lightsteelblue;
    padding: 10px;
}