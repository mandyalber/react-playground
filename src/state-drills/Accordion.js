import React from 'react';
import './Accordion.css'

export default class Accordion extends React.Component{
 //You can manage which section is currently active or opened by keeping track of the index of the last clicked button in state.
    state = {indexOfLastClicked: 0};
    static defaultProps = {sections: []};

    handleButtonClicked(index){
        this.setState({indexOfLastClicked: index})
    }

    renderButtons(){      
        return ( 
            this.props.sections.map((section, index)=>(
                <li key={index} className="section">
                    <button key={index} onClick={()=>this.handleButtonClicked(index)}>
                        {section.title}
                    </button>   
                    {(this.state.indexOfLastClicked === index) && this.renderContent()}
                </li>   
            ))    
        )       
    }     

    renderContent(){
        const currentSection = this.props.sections[this.state.indexOfLastClicked]
        return (
            <p className="content">{currentSection.content}</p>
        )
    }

    render (){
        return (
            <ul className="accordion">
                {this.renderButtons()}
            </ul>
        )
    }
}    

/*add styles for the Accordion component in an Accordion.css file.*/