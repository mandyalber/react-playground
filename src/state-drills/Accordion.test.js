import React from 'react';
import ReactDOM from 'react-dom';
import Accordion from './Accordion';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json'

const sectionsProp = [
    {
      title: 'Section 1',
      content: 'Lorem ipsum dolor sit amet consectetur adipisicing elit.',
    },
    {
      title: 'Section 2',
      content: 'Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!',
    },
    {
      title: 'Section 3',
      content: 'Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?',
    },
  ];

describe('Accordion component', ()=>{
    //smoke test
    it('renders without errors', ()=>{
        const div = document.createElement('div');
        ReactDOM.render(<Accordion />, div);
        ReactDOM.unmountComponentAtNode(div);
    })
    //snapshot tests
    it('renders an empty li with no sections prop', ()=>{
        const wrapper = shallow(<Accordion />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })
    it('renders no sections as active by default', ()=>{
        const wrapper = shallow(<Accordion sections={sectionsProp} />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })
    it('opens a clicked section', ()=>{
        const wrapper = shallow(<Accordion sections={sectionsProp} />)
        wrapper.find('button').at(1).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })
    it('only opens the last section', ()=>{
        const wrapper = shallow(<Accordion sections={sectionsProp} />)
        wrapper.find('button').at(1).simulate('click')
        wrapper.find('button').at(2).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })
})