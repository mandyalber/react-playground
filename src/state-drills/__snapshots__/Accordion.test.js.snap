// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component only opens the last section 1`] = `
<ul
  className="accordion"
>
  <li
    className="section"
    key="0"
  >
    <button
      key="0"
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li
    className="section"
    key="1"
  >
    <button
      key="1"
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    className="section"
    key="2"
  >
    <button
      key="2"
      onClick={[Function]}
    >
      Section 3
    </button>
    <p
      className="content"
    >
      Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?
    </p>
  </li>
</ul>
`;

exports[`Accordion component opens a clicked section 1`] = `
<ul
  className="accordion"
>
  <li
    className="section"
    key="0"
  >
    <button
      key="0"
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li
    className="section"
    key="1"
  >
    <button
      key="1"
      onClick={[Function]}
    >
      Section 2
    </button>
    <p
      className="content"
    >
      Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!
    </p>
  </li>
  <li
    className="section"
    key="2"
  >
    <button
      key="2"
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion component renders an empty li with no sections prop 1`] = `
<ul
  className="accordion"
/>
`;

exports[`Accordion component renders no sections as active by default 1`] = `
<ul
  className="accordion"
>
  <li
    className="section"
    key="0"
  >
    <button
      key="0"
      onClick={[Function]}
    >
      Section 1
    </button>
    <p
      className="content"
    >
      Lorem ipsum dolor sit amet consectetur adipisicing elit.
    </p>
  </li>
  <li
    className="section"
    key="1"
  >
    <button
      key="1"
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    className="section"
    key="2"
  >
    <button
      key="2"
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;
